src/app/common/FabricTableImpl.hpp

```diff
@@ -0,0 +1,835 @@
+{
+    using Serializer           = DefaultSerializer<StorageId, StorageData>;
+    using TypedFabricEntryData = FabricEntryData<StorageId, StorageData, Serializer::kPersistentStorageDataBufferMax(),
+                                                 Serializer::kPersistentFabricBufferMax(), Serializer::kMaxPerFabric()>;
+
+    VerifyOrReturnError(IsInitialized(), CHIP_ERROR_INTERNAL);
+
+    TypedFabricEntryData fabric(mEndpointId, fabric_index);
+    CHIP_ERROR err = fabric.Load(mStorage);
+    VerifyOrReturnError(CHIP_NO_ERROR == err || CHIP_ERROR_NOT_FOUND == err, err);
+
+    entry_count = (CHIP_ERROR_NOT_FOUND == err) ? 0 : fabric.entry_count;
+
+    return CHIP_NO_ERROR;
+}
+
+template <class StorageId, class StorageData, size_t kIteratorsMax>
+CHIP_ERROR FabricTableImpl<StorageId, StorageData, kIteratorsMax>::GetEndpointEntryCount(uint8_t & entry_count)
+{
+    using TypedEndpointEntryCount = EndpointEntryCount<StorageId, StorageData>;
+
+    VerifyOrReturnError(IsInitialized(), CHIP_ERROR_INTERNAL);
+
+    TypedEndpointEntryCount endpoint_entry_count(mEndpointId);
+
+    ReturnErrorOnFailure(endpoint_entry_count.Load(mStorage));
+    entry_count = endpoint_entry_count.count_value;
+
+    return CHIP_NO_ERROR;
+}
+
+template <class StorageId, class StorageData, size_t kIteratorsMax>
+CHIP_ERROR FabricTableImpl<StorageId, StorageData, kIteratorsMax>::SetEndpointEntryCount(const uint8_t & entry_count)
+{
+    using TypedEndpointEntryCount = EndpointEntryCount<StorageId, StorageData>;
+    VerifyOrReturnError(IsInitialized(), CHIP_ERROR_INTERNAL);
+
+    TypedEndpointEntryCount endpoint_entry_count(mEndpointId, entry_count);
+    return endpoint_entry_count.Save(mStorage);
+}
+
+template <class StorageId, class StorageData, size_t kIteratorsMax>
+CHIP_ERROR FabricTableImpl<StorageId, StorageData, kIteratorsMax>::GetRemainingCapacity(FabricIndex fabric_index,
+                                                                                        uint8_t & capacity)
+{
+    using Serializer           = DefaultSerializer<StorageId, StorageData>;
+    using TypedFabricEntryData = FabricEntryData<StorageId, StorageData, Serializer::kPersistentStorageDataBufferMax(),
+                                                 Serializer::kPersistentFabricBufferMax(), Serializer::kMaxPerFabric()>;
+
+    VerifyOrReturnError(IsInitialized(), CHIP_ERROR_INTERNAL);
+
+    uint8_t endpoint_entry_count = 0;
+    ReturnErrorOnFailure(GetEndpointEntryCount(endpoint_entry_count));
+
+    // If the global entry count is higher than the maximal Global entry capacity, this returns a capacity of 0 until enough entries
+    // have been deleted to bring the global number of entries under the global maximum.
+    if (endpoint_entry_count > mMaxPerEndpoint)
+    {
+        capacity = 0;
+        return CHIP_NO_ERROR;
+    }
+    uint8_t remaining_capacity_global = static_cast<uint8_t>(mMaxPerEndpoint - endpoint_entry_count);
```

Comment:
The fabric capacity vs global capacity and the way to handle capacity changes on OTA was specific to Scenes due to the Fabric Capacit = Half -1 rule. I think we should document this in the FabricTableImpl class if we want to keep or change the logic.

Reply:
Right, I suppose setting an appropriate `mMaxPerEndpoint` would ensure the behavior works as expected. I see the OTA unit test failing so that's why I thought that maybe some Scene Specific behavior was generalized here.

Reply:
All tests passing now